File: src/main/java/no/nav/veilarbaktivitet/veilarbportefolje/AktivitetKafkaProducerService.java
package no.nav.veilarbaktivitet.veilarbportefolje;

import io.micrometer.core.annotation.Timed;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import no.nav.common.utils.IdUtils;
import no.nav.veilarbaktivitet.aktivitet.AktivitetService;
import no.nav.veilarbaktivitet.aktivitet.domain.AktivitetData;
import no.nav.veilarbaktivitet.config.kafka.kafkatemplates.KafkaJsonTemplate;
import no.nav.veilarbaktivitet.config.kafka.kafkatemplates.KafkaStringTemplate;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.common.header.internals.RecordHeader;
import org.slf4j.MDC;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.kafka.support.SendResult;
import org.springframework.stereotype.Service;
import org.springframework.util.concurrent.ListenableFuture;

import static no.nav.common.kafka.producer.util.ProducerUtils.toJsonProducerRecord;
import static no.nav.common.rest.filter.LogRequestFilter.NAV_CALL_ID_HEADER_NAME;

@Slf4j
@RequiredArgsConstructor
@Service
public class AktivitetKafkaProducerService {

    private final KafkaStringTemplate portefoljeProducer;
    private final KafkaJsonTemplate<AktivitetData> aktivitetProducer;
    private final AktivitetService aktivitetService;
    private final KafkaAktivitetDAO dao;

    @Value("${topic.ut.portefolje}")
    private String portefoljeTopic;

    @Value("${topic.ut.aktivitetdata.rawjson}")
    private String aktivitetTopic;

    @Timed("aktivitet_til_kafka")
    @SneakyThrows
    public void sendAktivitetMelding(KafkaAktivitetMeldingV4 melding) {
        AktivitetData aktivitetData = aktivitetService.hentAktivitetMedFHOForVersion(melding.getVersion());
        ProducerRecord<String, String> portefoljeMelding = toJsonProducerRecord(portefoljeTopic, melding.getAktorId(), melding);
        portefoljeMelding.headers().add(new RecordHeader(NAV_CALL_ID_HEADER_NAME, getCorrelationId().getBytes()));

        ListenableFuture<SendResult<String, AktivitetData>> send = aktivitetProducer.send(aktivitetTopic, aktivitetData.getAktorId(), aktivitetData);
        long offset = portefoljeProducer.send(portefoljeMelding).get().getRecordMetadata().offset();
        send.get();

        dao.updateSendtPaKafkaAven(melding.getVersion(), offset);
    }

    static String getCorrelationId() {
        String correlationId = MDC.get(NAV_CALL_ID_HEADER_NAME);

        if (correlationId == null) {
            correlationId = MDC.get("jobId");
        }
        if (correlationId == null) {
            correlationId = IdUtils.generateId();
        }

        return correlationId;
    }

}

File: src/main/kotlin/no/nav/veilarbaktivitet/veilarbportefolje/AktivitetKafkaProducerService.kt
package no.nav.veilarbaktivitet.veilarbportefolje

import io.micrometer.core.annotation.Timed
import lombok.RequiredArgsConstructor
import lombok.SneakyThrows
import lombok.extern.slf4j.Slf4j
import no.nav.common.json.JsonUtils
import no.nav.common.rest.filter.LogRequestFilter
import no.nav.veilarbaktivitet.config.kafka.kafkatemplates.KafkaStringTemplate
import no.nav.veilarbaktivitet.veilarbportefolje.dto.KafkaAktivitetMeldingV4
import org.apache.kafka.clients.producer.ProducerRecord
import org.apache.kafka.common.header.internals.RecordHeader
import org.apache.kafka.common.header.internals.RecordHeaders
import org.slf4j.MDC
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import java.util.*

@Slf4j
@RequiredArgsConstructor
@Service
class AktivitetKafkaProducerService(
    private val portefoljeProducer: KafkaStringTemplate,
    private val dao: KafkaAktivitetDAO,
    @Value("\${topic.ut.portefolje}")
    private val portefoljeTopic: String
) {

    @Timed("aktivitet_til_kafka")
    @SneakyThrows
    fun sendAktivitetMelding(melding: KafkaAktivitetMeldingV4) {
        val portefoljeMelding = ProducerRecord(
            portefoljeTopic,
            null,
            melding.aktorId,
            JsonUtils.toJson(melding),
            RecordHeaders().add(
                RecordHeader(LogRequestFilter.NAV_CALL_ID_HEADER_NAME, correlationId.toByteArray())
            )
        )

        val offset = portefoljeProducer.send(portefoljeMelding).get().recordMetadata.offset()

        dao.updateSendtPaKafkaAven(melding.version, offset)
    }

    companion object {
        val correlationId: String
            get() {
                return MDC.get(LogRequestFilter.NAV_CALL_ID_HEADER_NAME)
                    ?: MDC.get("jobId")
                    ?: UUID.randomUUID().toString()
            }
    }
}
